callbacks.go

```diff
@@ -68,35 +68,60 @@ func getMethod(value reflect.Value, name string) reflect.Value {
 	return method
 }
 
-// Get methods from the given value and any embedded fields.
+// getMethods gets all methods with the given name from the given value
+// and any embedded fields.
+//
+// Returns a slice of bound methods that can be called directly.
 func getMethods(value reflect.Value, name string) []reflect.Value {
-	// Collect all possible receivers
 	receivers := []reflect.Value{value}
-	if value.Kind() == reflect.Ptr {
-		value = value.Elem()
-	}
-	if value.Kind() == reflect.Struct {
-		t := value.Type()
-		for i := 0; i < value.NumField(); i++ {
-			field := value.Field(i)
-			fieldType := t.Field(i)
-			if !fieldType.IsExported() {
-				continue
-			}
 
-			// Hooks on exported embedded fields should be called.
-			if fieldType.Anonymous {
-				receivers = append(receivers, field)
-				continue
-			}
+	// Traverse all embedded fields of the struct
+	// starting from the given value
+	// and collect all possible receivers.
+	//
+	// Two kinds of embedded fields are considered if they're exported:
+	//
+	//   - standard Go embedded fields
+	//   - fields tagged with `embed:""`
+	for pending := []reflect.Value{value}; len(pending) > 0; pending = pending[1:] {
```

Comment:
I think this implementation would be much much simpler using recursion.

Reply:
Heh, I did a recursive one first and switched to this. I'll switch it back. Give me a couple minutes.

Reply:
Updated: simplified the management of the 'receivers' slice as well so the same conditions don't need to be checked multiple times.